celery_amqp_backend/backend.py

```diff
@@ -379,3 +381,259 @@ def __reduce__(self, args=(), kwargs=None):
+            pass
+
+        if meta:
+            consumer = self._consumers.pop(task_id, None)
+            if consumer and consumer not in self._consumers.values():
+                consumer.cancel()
+
+            return self.meta_from_decoded(meta)
+        else:
+            return {
+                'status': states.PENDING,
+                'result': None,
+            }
+
+    def store_result(self, task_id, result, state, traceback=None, request=None, **kwargs):
+        """
+        Sends the task result for the given task identifier to the task result queue and returns the sent result dict.
+
+        :param task_id: Task identifier to send the result for
+        :param result: The task result as dict
+        :param state: The task result state
+        :param traceback: The traceback if the task resulted in an exception
+        :param request: Request data
+        :param kwargs:
+        :return: The task result as dict
+        """
+        # Determine the routing key and a potential correlation identifier.
+        routing_key = self._create_routing_key(task_id, request)
+        correlation_id = self._create_correlation_id(task_id, request)
+
+        with self.app.amqp.producer_pool.acquire(block=True) as producer:
+            producer.publish(
+                {
+                    'task_id': task_id,
+                    'status': state,
+                    'result': self.encode_result(result, state),
+                    'traceback': traceback,
+                    'children': self.current_task_children(request),
+                },
+                exchange='',
+                routing_key=routing_key,
+                correlation_id=correlation_id,
+                serializer=self.serializer,
+                retry=True,
+                retry_policy=self.retry_policy,
+                delivery_mode=self.delivery_mode,
+            )
+
+        return result
+
+    def on_task_call(self, producer, task_id):
+        """
+        Creates and saves a consumer for the direct-reply pseudo-queue, before the task request is sent
+        to the queue.
+
+        :param producer: The producer for the task request
+        :param task_id: The task identifier
+        """
+        for _, consumer in self._consumers.items():
+            if consumer.channel is producer.channel:
+                self._consumers[task_id] = consumer
+                break
+        else:
+            self._consumers[task_id] = self._create_consumer(
+                producer.channel,
+            )
+
+    def _create_consumer(self, channel):
+        """
+        Creates a consumer with the given parameters.
+
+        :param channel: The channel to use for the consumer
+        :return: Created consumer
+        """
+        consumer_queue = kombu.Queue("amq.rabbitmq.reply-to", no_ack=True)
+        consumer = kombu.Consumer(
+            channel,
+            queues=[consumer_queue],
+            auto_declare=True,
+        )
```

Comment:
This call fails when multiple concurrent executions happen.
Exception: `ConcurrentObjectUseError`
Message: `This socket is already used by another greenlet: <bound method Waiter.switch of <gevent._gevent_c_waiter.Waiter object at 0x7fb6c0c1ebf0>>`
Traceback: 
```
  File \"/usr/local/lib/python3.7/site-packages/celery/app/base.py\", line 787, in send_task
    self.backend.on_task_call(P, task_id)
  File \"/app/app/celery_amqp_backend/backend.py\", line 576, in on_task_call
    producer.channel,
  File \"/app/app/celery_amqp_backend/backend.py\", line 590, in _create_consumer
    auto_declare=True,
  File \"/usr/local/lib/python3.7/site-packages/kombu/messaging.py\", line 387, in __init__
    self.revive(self.channel)
  File \"/usr/local/lib/python3.7/site-packages/kombu/messaging.py\", line 409, in revive
    self.declare()
  File \"/usr/local/lib/python3.7/site-packages/kombu/messaging.py\", line 422, in declare
    queue.declare()
  File \"/usr/local/lib/python3.7/site-packages/kombu/entity.py\", line 606, in declare
    self._create_queue(nowait=nowait, channel=channel)
  File \"/usr/local/lib/python3.7/site-packages/kombu/entity.py\", line 615, in _create_queue
    self.queue_declare(nowait=nowait, passive=False, channel=channel)
  File \"/usr/local/lib/python3.7/site-packages/kombu/entity.py\", line 650, in queue_declare
    nowait=nowait,
  File \"/usr/local/lib/python3.7/site-packages/amqp/channel.py\", line 1163, in queue_declare
    spec.Queue.DeclareOk, returns_tuple=True,
  File \"/usr/local/lib/python3.7/site-packages/amqp/abstract_channel.py\", line 99, in wait
    self.connection.drain_events(timeout=timeout)
  File \"/usr/local/lib/python3.7/site-packages/amqp/connection.py\", line 525, in drain_events
    while not self.blocking_read(timeout):
  File \"/usr/local/lib/python3.7/site-packages/amqp/connection.py\", line 530, in blocking_read
    frame = self.transport.read_frame()
  File \"/usr/local/lib/python3.7/site-packages/amqp/transport.py\", line 294, in read_frame
    frame_header = read(7, True)
  File \"/usr/local/lib/python3.7/site-packages/amqp/transport.py\", line 627, in _read
    s = recv(n - len(rbuf))
  File \"/usr/local/lib/python3.7/site-packages/gevent/_socketcommon.py\", line 663, in recv
    self._wait(self._read_event)
  File \"src/gevent/_hub_primitives.py\", line 317, in gevent._gevent_c_hub_primitives.wait_on_socket
  File \"src/gevent/_hub_primitives.py\", line 322, in gevent._gevent_c_hub_primitives.wait_on_socket
  File \"src/gevent/_hub_primitives.py\", line 297, in gevent._gevent_c_hub_primitives._primitive_wait
```